File: digital_wallet_php/src/Service/WalletService.php
<?php

namespace App\Service;

use App\Config\Database;
use App\Exception\WalletNotFoundException;
use App\Util\Timestamp;

class WalletService
{
    public function getWalletByUserId(int $userId): array
    {
        $db = Database::getConnection();

        $stmt = $db->prepare('SELECT * FROM wallets WHERE user_id = :user_id');
        $stmt->execute([':user_id' => $userId]);
        $wallet = $stmt->fetch();

        if (!$wallet) {
            throw new WalletNotFoundException("Wallet not found for userId: {$userId}");
        }

        return [
            'id' => (int) $wallet['id'],
            'userId' => (int) $wallet['user_id'],
            'currency' => $wallet['currency'],
            'balance' => round((float) $wallet['balance'], 4),
            'version' => (int) $wallet['version'],
            'updatedAt' => Timestamp::format($wallet['updated_at']),
        ];
    }
}

File: digital_wallet_springmvc/src/main/java/com/digitalwallet/service/WalletService.java
package com.digitalwallet.service;

import com.digitalwallet.exception.WalletNotFoundException;
import com.digitalwallet.mapper.WalletMapper;
import com.digitalwallet.model.Wallet;
import com.digitalwallet.model.WalletDTO;

public class WalletService {

    private WalletMapper walletMapper;

    public void setWalletMapper(WalletMapper walletMapper) { this.walletMapper = walletMapper; }

    public WalletDTO getByUserId(Long userId) {
        Wallet wallet = walletMapper.findByUserId(userId);
        if (wallet == null) {
            throw new WalletNotFoundException("Wallet not found for userId: " + userId);
        }

        return WalletDTO.builder()
                .id(wallet.getId())
                .userId(wallet.getUserId())
                .currency(wallet.getCurrency())
                .balance(wallet.getBalance())
                .version(wallet.getVersion())
                .updatedAt(wallet.getUpdatedAt())
                .build();
    }
}
